### ISSUE TITLE
feat(graph): Add autograd engine for learnable ROI weights (Phase 2)

### ISSUE THREAD

**Issue by Dongckim** [Dongckim/cortex-sdk#2]
## Overview

Add `cortex/graph/autograd.py` implementing reverse-mode autodiff (micrograd-style) so the ROI scoring weights become learnable parameters.

## Background

The current ROI scoring formula in `cortex/optimizer/hybrid_roi.py`:

```
S = wc*Sc + wt*St + ws*Ss_adj + wm*Sm
```

The weights `(wc, wt, ws, wm)` are currently hardcoded per `RequestType`. This phase makes them differentiable and tunable via gradient descent.

### `autograd.py`
- `Value` class wrapping scalar ops (`add`, `mul`, `relu`, `sigmoid`) that tracks computation history
- `.backward()` implementing reverse-mode autodiff via chain rule
- Weights in `HybridROI` scorer become `Value` instances (learnable parameters)

### `examples/tune_weights.py`
A gradient descent training loop that:
1. Takes a small labeled dataset: `(image, ideal_crop_region)` pairs
2. Runs forward pass, computes IoU loss between predicted and ideal crop
3. Backprops and updates weights for N steps
4. Prints loss curve

### Additional Compiler Passes
- **`common_subexpression_elimination`**: detect nodes computing identical ops and deduplicate
- **`constant_folding`**: fold `wc*Sc` when `Sc` is a known constant tensor

## Notes
- Use mathematical notation in docstrings (chain rule, Jacobian)
- Builds on top of the Graph IR from Phase 1 (issue #1)

## Files to Read First
- `cortex/optimizer/hybrid_roi.py`

**Comment by Dongckim** [comment]
## Closing Phase 2 — Autograd / Weight Learning

After deeper analysis of the cortex L2 pipeline, this phase is not viable as designed. Here's why.

### What the plan was

Learn the score fusion weights (wc, wt, ws, wm) via gradient descent:

```
S = wc*Sc + wt*St + ws*Ss_adj + wm*Sm
loss = MSE(S, ideal_S)
backward() → update wc, wt, ws, wm
```

### Why it doesn't work

**1. argmax is not differentiable**

ROI selection picks cells via argmax over the score map — a discrete operation. A small change in wc produces discontinuous jumps in the selected region, so `∂loss/∂wc` is undefined. The backward path breaks here.

**2. No ground-truth data exists**

Gradient descent requires `ideal_S`, but there is no labeled dataset of `(image, ideal_crop_region)` pairs. Synthetic targets are meaningless; VLM feedback as supervision adds too many moving parts.

**3. MSER as a supervision signal doesn't save it**

Using MSER output as fixed ground-truth (outside the backward path) sidesteps the differentiability issue for text detection. But the argmax problem in crop selection remains — the gradient path is still broken.

### Root cause

Cortex is not a learning-based system. There is no continuous, argmax-free loss function that can be defined over its discrete crop selection. Forcing an autograd layer on top would produce code that runs but doesn't actually learn anything.

### Decision

Closing this issue. Development continues with Phase 1 (Graph IR + lowering passes) and Phase 3 (operator fusion + benchmark), which fit naturally with the existing codebase structure.
